src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs

```diff
@@ -0,0 +1,353 @@
+                }
+
+                if (ipv4Address is null)
+                {
+                    Debug.Assert(isVersion4a);
+                    buffer[4] = 0;
+                    buffer[5] = 0;
+                    buffer[6] = 0;
+                    buffer[7] = 255;
+                }
+                else
+                {
+                    ipv4Address.TryWriteBytes(buffer.AsSpan(4), out int bytesWritten);
+                    Debug.Assert(bytesWritten == 4);
+                }
+
+                byte usernameLength = checked((byte)Encoding.UTF8.GetBytes(username, buffer.AsSpan(8)));
+                buffer[8 + usernameLength] = 0;
+                int totalLength = 9 + usernameLength;
+
+                if (ipv4Address is null)
+                {
+                    // https://www.openssh.com/txt/socks4a.protocol
+                    byte hostLength = checked((byte)Encoding.UTF8.GetBytes(host, buffer.AsSpan(totalLength)));
+                    buffer[totalLength + hostLength] = 0;
+                    totalLength += hostLength + 1;
+                }
+
+                await WriteAsync(stream, buffer.AsMemory(0, totalLength), async).ConfigureAwait(false);
+
+                // +----+----+----+----+----+----+----+----+
+                // | VN | CD | DSTPORT |      DSTIP        |
+                // +----+----+----+----+----+----+----+----+
+                //    1    1      2              4
+                await ReadToFillAsync(stream, buffer.AsMemory(0, 8), async).ConfigureAwait(false);
+                VerifyProtocolVersion(ProtocolVersion4, buffer[0]);
+
+                switch (buffer[1])
+                {
+                    case Socks4_Success:
+                        // Nothing to do
+                        break;
+                    case Socks4_AuthFailed:
+                        throw new SocksException(SR.net_socks_auth_failed);
+                    default:
+                        throw new SocksException(SR.net_socks_connection_failed);
+                }
+                // response address not used
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
+        }
+
+        private static void VerifyProtocolVersion(byte expected, byte version)
+        {
+            if (expected != version)
+            {
+                throw new SocksException(SR.Format(SR.net_socks_unexpected_version, expected, version));
+            }
+        }
+
+        private static ValueTask WriteAsync(Stream stream, Memory<byte> buffer, bool async)
+        {
+            if (async)
+            {
+                return stream.WriteAsync(buffer);
+            }
+            else
+            {
+                stream.Write(buffer.Span);
+                return default;
+            }
+        }
+
+        private static async ValueTask ReadToFillAsync(Stream stream, Memory<byte> buffer, bool async)
+        {
+            while (!buffer.IsEmpty)
```

Comment:
We prefer to use `Length`, to avoid any confusion of 0-length memory vs null pointer.

```suggestion
            while (buffer.Length != 0)
```